content/posts/jpa_spring_data_query_optimisations.en.md

```diff
@@ -0,0 +1,213 @@
+}
+```
+
+Hibernate will query the database in this way:
+* 1 query to select the main entity
+* N queries for the entities linked by  the jointure
+
+In our case we can see the following queries in the logs:
+
+```shell
+Hibernate: select s1_0.id,s1_0.name from store s1_0
+Hibernate: select b1_0.store_id,b1_0.id,b1_0.description,b1_0.isbn_10,b1_0.isbn_13,b1_0.medium_image_url,b1_0.nb_of_pages,b1_0.price,b1_0.rank,b1_0.small_image_url,b1_0.title,b1_0.year_of_publication from book b1_0 where b1_0.store_id=?
+Hibernate: select a1_0.books_id,a1_1.id,a1_1.firstname,a1_1.lastname,a1_1.public_id from book_authors a1_0 join author a1_1 on a1_1.id=a1_0.authors_id where a1_0.books_id=?
+Hibernate: select a1_0.books_id,a1_1.id,a1_1.firstname,a1_1.lastname,a1_1.public_id from book_authors a1_0 join author a1_1 on a1_1.id=a1_0.authors_id where a1_0.books_id=?
+Hibernate: select a1_0.books_id,a1_1.id,a1_1.firstname,a1_1.lastname,a1_1.public_id from book_authors a1_0 join author a1_1 on a1_1.id=a1_0.authors_id where a1_0.books_id=?
+Hibernate: select a1_0.books_id,a1_1.id,a1_1.firstname,a1_1.lastname,a1_1.public_id from book_authors a1_0 join author a1_1 on a1_1.id=a1_0.authors_id where a1_0.books_id=?
+[...]
+```
+
+It's unfortunately not finished, 
+
+Imagine now, your book entity is related to another one in a EAGER way.
+
+```java
+@ManyToMany(fetch = FetchType.EAGER)
+private List<Author> authors;
+```
+
+You will execute then another SQL queries.
+
+For instance, in this case:
+
+```jshelllanguage
+Hibernate: select b1_0.authors_id,b1_1.id,b1_1.description,b1_1.isbn_10,b1_1.isbn_13,b1_1.medium_image_url,b1_1.nb_of_pages,b1_1.price,b1_1.rank,b1_1.small_image_url,s1_0.id,s1_0.name,b1_1.title,b1_1.year_of_publication from book_authors b1_0 join book b1_1 on b1_1.id=b1_0.books_id left join store s1_0 on s1_0.id=b1_1.store_id where b1_0.authors_id=?
+```
+
+{{< admonition tip "To sum up" true >}}
+At the same way SQL jointures are really time-consuming, the way you can link entities may strongly impact the performance of your queries in either memory or while running SQL queries.   
+{{< /admonition >}}
+
+## Control the number of EAGER/SELECT relations underlying queries using ``@BatchSize``
+We can easily reduce the number of SELECT queries while fetching another entities with the ``@BatchSize`` annotation
+
+```java
+@Entity
+public class Store{
+[...]
+    @OneToMany(fetch = FetchType.EAGER,mappedBy = "store")
+    @BatchSize(size = 5)
+    private List<Book> books;
+[...]
+```
+
+TODO LOGS
+
+### Use a dedicated entity graph
+If you are still struggling with the way Hibernate loads your Entity graph, you can also try to specify the graph of entities to load by yourself.
+It could be really useful if you want to avoid to retrieve specific useless attributes which make your queries really slow.
+
+[JPA 2.1 has introduced this feature](https://docs.oracle.com/javaee/7/tutorial/persistence-entitygraphs001.htm).
```

Comment:
Maybe there should not be a link to Java EE 7 anymore; the best I could find for Jakarta JPA is https://jakarta.ee/learn/docs/jakartaee-tutorial/current/persist/persistence-entitygraphs/persistence-entitygraphs.html